### PR TITLE
Use declarative metadata with `pyproject.toml`

### PR DESCRIPTION
**Pull request by AA-Turner**
Requires #10355, #10352 

This moves Sphinx's package metadata to the standards-compliant format, preserving setuptools-specific items with the `tool.setuptools` backend.

### Relates
- https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

### CODE REVIEW

**Comment by AA-Turner** [comment]
It seems that `tox` requires fully building and installing the package, and `setuptools` dropped support for Python 3.6 in December 2021 (https://setuptools.pypa.io/en/latest/history.html#v59-7-0), before the support for `pyproject.toml` was added.

I'm going to mark this as draft for now, I may look at e.g. flit later (that Sphinx uses `egg_info` configuration slightly complicates things, though).

A

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,123 @@
+[build-system]
+requires = ["setuptools>=62", "wheel"]
```
Comment (tk0miya) [inline pyproject.toml]: -1: I think this change also affect to users. This requires users who try to install Sphinx via source package must upgrade setuptools to the newer one. For example, the HEAD of Ubuntu still bundles 44.1.
https://packages.ubuntu.com/impish/python-setuptools

Comment (AA-Turner) [reply pyproject.toml]: > This requires users who try to install Sphinx via source package must upgrade setuptools to the newer one.

Local builds (e.g. `pip install .`) now default to isolated environments, which download the build requirements from PyPI, so this wouldn't be an issue, per se.

The bigger issue is that setuptools 62 does not support Python 3.6 at all -- I could re-target this PR to master, but we would then have two different build systems for 12 months (whilst the 5.x branch is active, still supporting Python 3.6) -- not a good solution.

A

**Comment by AA-Turner** [comment]
It was easier than expected to move to `flit`, so I'm un-marking as draft. CI is passing.

A

**Comment by nanonyme** [comment]
To be fair, Python 3.6 is since EOL this January.

**Comment by tk0miya** [comment]
As I commented at https://github.com/sphinx-doc/sphinx/issues/9820#issuecomment-1140500002, my large concern is supporting users. I suppose most of them would not know python at all. So I hesitate to depend on the latest libraries.

I can accept to migrate package metadata that is not related to install itself (ex. classifiers). But I'm not sure it's worthy for us to migrate it partially.

**Comment by nanonyme** [comment]
Users using pip will not notice any difference with this.

**Comment by AA-Turner** [comment]
> As I commented at https://github.com/sphinx-doc/sphinx/issues/9820#issuecomment-1140500002, my large concern is supporting users. I suppose most of them would not know python at all. So I hesitate to depend on the latest libraries.

I entirely agree we should be supporting users. I see three groups:

1. "python user" -- uses `pip install sphinx`
2. "distro user" -- uses `apt install sphinx` (or yum, dnf, etc.)
3. "sources user" -- uses `git clone https://github.com/sphinx-doc/sphinx.git && cd sphinx && python setup.py build` or similar

For group 1, `pip` manages everything automatically. It fetches pacakges matching the constraints that we specify, and ensures a consistent installation.

For group 2, the redistributor (debian, RHEL, ubuntu, etc) ensures that the system has a consistent installation, and installs everything themselves.

For group 3, there are no guarantees at all -- it is up to the user to ensure every dependency is installed and there are no conflicts.

I would argue that the users of Sphinx are mostly in group 1, then group 2, then 3. There are [five million](https://pypistats.org/packages/sphinx) downloads of Sphinx per month from PyPI. We can't know numbers for redistributors, or proxies of PyPI, so this is an inexact science.

Group 2 users are in a precarious position when they attempt to use extensions like `autodoc` -- as Sphinx needs to import the module being documented, all the module's dependencies must be present. This might create conflicts with other Python packages in the distribution's global installation, which is why environments of some kind (`venv`, `virtualenv`, `conda` env) are so strongly reccomended.

I would like to solve these challenges -- having a single distributable self-contained binary that is "sphinx" would be amazing. It would free us from constraints on system Python versions etc, whilst allowing integration with a particular python version for those that install via `pip`. However we don't have this (sadly).

My conclusions are the following:
- `pip` users will be fine with any sort of version constraints, as `pip` is a proper package manager.
- users of distribution (ubuntu etc) supplied Sphinx binaries will be fine, as distributions manage all their versions independently of us, often pinning more strictly etc to ensure that the system as a whole has a consistent Python installation.
- Hence, *we should not worry about redistributors*.
- Users who install from source are assumed to be experts, and we do not provide any extra support here.

---------

The issue in this PR -- changing our build process -- is unrelated to any of these issues. For users of `pip`, pre-build wheels are downloaded from PyPI. For distribution users, only the packaging team need to care about our build process, and all distributions have PEP 517 / 518 tool chains now. From this PR, **nothing will change for end users**.

---------

I'm happy to propose a PR to Sphinx's policies @tk0miya to outline this and provide a place for more discussion, but I would strongly argue that we can make these important improvements and changes in our build system and dependencies *without negatively affecting end users*, which is the most important thing.

A

(Sorry for the long message, I didn't intend to write this much!)

**Comment by nanonyme** [comment]
Note in case 3 you can do `pip install .` and it becomes case 1 again.

**Comment by danieleades** [comment]
mypy config should also be moved into `pyproject.toml` file

**Comment by danieleades** [comment]
as an aside, is there any interest in moving to a full-blown package manager, such as [Poetry](https://python-poetry.org/)?

**Comment by AA-Turner** [comment]
I would like to move forwards with this--@tk0miya does my [earlier comment](https://github.com/sphinx-doc/sphinx/pull/10356#issuecomment-1140534724) make sense?

The one line summary is that I propose we only support users installing from wheels on PyPI -- anything else is non-standard, and as we are volunteers here I think reasonable to put those limits on it.

A

**Comment by AA-Turner** [comment]
I've gone ahead with this for 5.2--though if we get reports from users that the new configuration seriously breaks things I will look at reverting the change.

A

File: CHANGES [hunk 2]
```diff
@@ -4,6 +4,10 @@ Release 5.2.0 (in development)
 Dependencies
 ------------
 
+* #10356: Sphinx now uses declarative metadata with ``pyproject.toml`` to
+  create packages, using PyPA's ``build`` project as a build backend. Patch by
```
Comment (mgorny) [inline CHANGES:8]: Just noticed this in the ChangeLog — `build` is actually a frontend, `flit_core` is the backend.

Comment (AA-Turner) [reply CHANGES:8]: Fixed, thank you.

A

Comment (mgorny) [reply CHANGES:8]: Thanks!

**Pull request merged**